tsdb/head_append_v2.go

```diff
@@ -0,0 +1,405 @@
+	if fh != nil {
+		if err := fh.Validate(); err != nil {
+			return 0, err
+		}
+	}
+
+	s := a.head.series.getByID(chunks.HeadSeriesRef(ref))
+	if s == nil {
+		var err error
+		s, _, err = a.getOrCreate(ls)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// TODO(PROM-60): Handle ST natively.
+	if a.head.opts.EnableSTAsZeroSample && st != 0 {
+		a.bestEffortAppendSTZeroSample(s, st, t, h, fh)
+	}
+
+	switch {
+	// Process the float histogram first as per the interface definition.
+	case fh != nil:
+		if err := a.appendFloatHistogram(s, t, fh, opts.RejectOutOfOrder); err != nil {
+			switch {
+			case errors.Is(err, storage.ErrOutOfOrderSample):
+				a.head.metrics.outOfOrderSamples.WithLabelValues(sampleMetricTypeHistogram).Inc()
+			case errors.Is(err, storage.ErrTooOldSample):
+				a.head.metrics.tooOldSamples.WithLabelValues(sampleMetricTypeHistogram).Inc()
+			}
+			return 0, err
+		}
+	case h != nil:
+		if err := a.appendHistogram(s, t, h, opts.RejectOutOfOrder); err != nil {
+			switch {
+			case errors.Is(err, storage.ErrOutOfOrderSample):
+				a.head.metrics.outOfOrderSamples.WithLabelValues(sampleMetricTypeHistogram).Inc()
+			case errors.Is(err, storage.ErrTooOldSample):
+				a.head.metrics.tooOldSamples.WithLabelValues(sampleMetricTypeHistogram).Inc()
+			}
+			return 0, err
+		}
+	default:
+		if value.IsStaleNaN(v) {
+			// If we have added a sample before with this same appender, we
+			// can check the previously used type and turn a stale float
+			// sample into a stale histogram sample or stale float histogram
+			// sample as appropriate. This prevents an unnecessary creation
+			// of a new batch. However, since other appenders might append
+			// to the same series concurrently, this is not perfect but just
+			// an optimization for the more likely case.
+			switch a.typesInBatch[s.ref] {
+			case stHistogram, stCustomBucketHistogram:
+				return a.Append(ref, ls, st, t, 0, &histogram.Histogram{Sum: v}, nil, storage.AOptions{
+					RejectOutOfOrder: opts.RejectOutOfOrder,
+				})
+			case stFloatHistogram, stCustomBucketFloatHistogram:
+				return a.Append(ref, ls, st, t, 0, nil, &histogram.FloatHistogram{Sum: v}, storage.AOptions{
+					RejectOutOfOrder: opts.RejectOutOfOrder,
+				})
+			}
+			// Note that a series reference not yet in the map will come out
+			// as stNone, but since we do not handle that case separately,
+			// we do not need to check for the difference between "unknown
+			// series" and "known series with stNone".
+		}
+		if err := a.appendFloat(s, t, v, opts.RejectOutOfOrder); err != nil {
+			switch {
+			case errors.Is(err, storage.ErrOutOfOrderSample):
+				a.head.metrics.outOfOrderSamples.WithLabelValues(sampleMetricTypeFloat).Inc()
+			case errors.Is(err, storage.ErrTooOldSample):
+				a.head.metrics.tooOldSamples.WithLabelValues(sampleMetricTypeFloat).Inc()
+			}
+			return 0, err
+		}
+	}
+
+	if t < a.mint {
+		a.mint = t
+	}
+	if t > a.maxt {
+		a.maxt = t
+	}
+
+	if value.IsStaleNaN(v) {
```

Comment:
Histogram can indicate staleness as well https://prometheus.io/docs/specs/native_histograms/#staleness-markers

```suggestion
	if (fh != nil && value.IsStaleNaN(fh.Sum)) || (h != nil && value.IsStaleNaN(h.Sum)) || value.IsStaleNaN(v) {
```

I'm not sure which unit test covers this, but probably needs cases for histograms.

Reply:
Added TODO on `TestMetadataInWAL_AppenderV2` hope that's ok for now? We never test stale sample appends + metadata.

> // TODO(bwplotka): Add cases ensuring stale sample appends will skipp metadata persisting.